handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslServerContext.java

```diff
@@ -121,9 +121,16 @@ static ServerContext newSessionContext(ReferenceCountedOpenSslContext thiz, long
                     // javadocs state that keyManagerFactory has precedent over keyCertChain, and we must have a
                     // keyManagerFactory for the server so build one if it is not specified.
                     if (keyManagerFactory == null) {
-                        keyMaterialProvider = new OpenSslCachingKeyMaterialProvider(
-                                chooseX509KeyManager(buildKeyManagerFactory(keyCertChain, key, keyPassword, null)
-                                        .getKeyManagers()), keyPassword);
+                        char[] keyPasswordChars = keyStorePassword(keyPassword);
+                        KeyStore ks = buildKeyStore(keyCertChain, key, keyPasswordChars);
+                        if (ks.aliases().hasMoreElements()) {
+                            keyManagerFactory = new OpenSslX509KeyManagerFactory();
+                        } else {
+                            keyManagerFactory = new OpenSslCachingX509KeyManagerFactory(
+                                    KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm()));
+                        }
+                        keyManagerFactory.init(ks, keyPasswordChars);
+                        keyMaterialProvider = providerFor(keyManagerFactory, keyPassword);
```

Comment:
This is now a duplicate of the else so it should be able to be pulled out of the branchs.